File: 322.零钱兑换.py
#
# @lc app=leetcode.cn id=322 lang=python3
#
# [322] 零钱兑换
#

# @lc code=start
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        self.dp = []
        if len(self.dp) <= amount:
            self.dp = [0 for _ in range(amount+1)]
        return self.change(coins, amount)

    def change(self, coins, amount):
        if amount < 0:
            return -1
        if amount == 0:
            return 0
        if (self.dp[amount] != 0): 
            return self.dp[amount] # 注意这个判断不能进下面的循环，因为存在下标为负的情况
        tol_num = -1
        for _, i in enumerate(coins):
            cur_num = self.change(coins, amount-i)+1 if self.change(coins, amount-i)>=0 else -1
            tol_num = min(cur_num, tol_num) if (cur_num>0 and tol_num>0) else max(cur_num, tol_num)
        self.dp[amount] = tol_num
        return tol_num
        
        
    '''
    正确但超时:
        Testcase [1,2,5]\n100， 
        expect 20
    '''
    # def change(self, coins, amount):
    #     if amount < 0:
    #         return -1
    #     if amount == 0:
    #         return 0
    #     tol_num = -1
    #     for _, i in enumerate(coins):
    #         cur_num = self.change(coins, amount-i)+1 if self.change(coins, amount-i)>=0 else -1
    #         tol_num = min(cur_num, tol_num) if (cur_num>0 and tol_num>0) else max(cur_num, tol_num)
    #     return tol_num
            

        
# @lc code=end

File: 322.零钱兑换.cpp
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        return changCoins(amount, coins, amount);
    }

    int changCoins(int amount, vector<int>& coins, int cur_amout) {
        if (cur_amout < 0 ) return -1;
        if (cur_amout == 0) return 0;

        int nums = INT_MAX;
        for (int coin: coins) {
            // cout<<nums<<endl;
            int sub = changCoins(amount, coins, cur_amout-coin);
            if (sub == -1) continue; // 减少不满足的条件中0的影响
            // int sub = changCoins(amount, coins, cur_amout-coin) + 1;
            // if (sub == 0) continue; // 减少不满足的条件中0的影响
            nums = min(sub + 1, nums); // 题目要求的是至少
            // cout<<coin<<": "<<nums<<endl;
        }
        return nums = INT_MAX? -1: nums;
    }
};
// @lc code=end

